File: packages/akribes-sdk-ts/src/sub/evals.ts
import type { HttpClient } from '../http';
import { nullOn404 } from '../http';
import type { EvalSuite, EvalSuiteSummary, EvalRun, EvalResult } from '../types';

export class EvalsClient {
  constructor(
    private http: HttpClient,
    private projectId: number,
  ) {}

  private path(scriptName: string, ...segments: string[]) {
    return this.http.scriptPath(this.projectId, scriptName, ...segments);
  }

  // ── Suites ────────────────────────────────────────────────────────────────

  async listSuites(scriptName: string, opts?: { signal?: AbortSignal }): Promise<EvalSuite[]> {
    return (await this.http.fetchOk(this.path(scriptName, 'eval-suites'), opts)).json();
  }

  async createSuite(
    scriptName: string,
    data: { name: string; runner_url: string; config?: Record<string, unknown>; auto_run_channels?: string[] },
    opts?: { signal?: AbortSignal },
  ): Promise<EvalSuite> {
    return (await this.http.fetchOk(this.path(scriptName, 'eval-suites'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
      signal: opts?.signal,
    })).json();
  }

  async getSuite(scriptName: string, suiteId: number, opts?: { signal?: AbortSignal }): Promise<EvalSuite | null> {
    return nullOn404(async () =>
      (await this.http.fetchOk(this.path(scriptName, 'eval-suites', String(suiteId)), opts)).json()
    );
  }

  async updateSuite(
    scriptName: string,
    suiteId: number,
    data: { runner_url?: string; config?: Record<string, unknown>; auto_run_channels?: string[] },
    opts?: { signal?: AbortSignal },
  ): Promise<EvalSuite> {
    return (await this.http.fetchOk(this.path(scriptName, 'eval-suites', String(suiteId)), {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
      signal: opts?.signal,
    })).json();
  }

  async deleteSuite(scriptName: string, suiteId: number, opts?: { signal?: AbortSignal }): Promise<void> {
    await this.http.fetchOk(this.path(scriptName, 'eval-suites', String(suiteId)), {
      method: 'DELETE',
      signal: opts?.signal,
    });
  }

  async checkRunnerHealth(
    scriptName: string,
    suiteId: number,
    opts?: { signal?: AbortSignal },
  ): Promise<{ status: string; http_status?: number; error?: string } | null> {
    return nullOn404(async () =>
      (await this.http.fetchOk(this.path(scriptName, 'eval-suites', String(suiteId), 'health'), opts)).json()
    );
  }

  // ── Trigger & Cancel ──────────────────────────────────────────────────────

  async trigger(
    scriptName: string,
    suiteId: number,
    data?: { source?: string; channel?: string; auto_publish?: boolean; triggered_by?: string },
    opts?: { signal?: AbortSignal },
  ): Promise<EvalRun> {
    return (await this.http.fetchOk(this.path(scriptName, 'eval-suites', String(suiteId), 'trigger'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data ?? {}),
      signal: opts?.signal,
    })).json();
  }

  async cancel(runId: number, opts?: { signal?: AbortSignal }): Promise<EvalRun> {
    return (await this.http.fetchOk(`${this.http.getBaseUrl()}/eval-runs/${runId}`, {
      method: 'DELETE',
      signal: opts?.signal,
    })).json();
  }

  // ── Runs & Results ────────────────────────────────────────────────────────

  async listRuns(
    scriptName: string,
    params?: { suite_id?: number; limit?: number; offset?: number },
    opts?: { signal?: AbortSignal },
  ): Promise<EvalRun[]> {
    const qs = new URLSearchParams();
    if (params?.suite_id != null) qs.set('suite_id', String(params.suite_id));
    if (params?.limit != null) qs.set('limit', String(params.limit));
    if (params?.offset != null) qs.set('offset', String(params.offset));
    const suffix = qs.toString() ? `?${qs}` : '';
    return (await this.http.fetchOk(this.path(scriptName, 'eval-runs') + suffix, opts)).json();
  }

  async getRun(runId: number, opts?: { signal?: AbortSignal }): Promise<EvalRun | null> {
    return nullOn404(async () =>
      (await this.http.fetchOk(`${this.http.getBaseUrl()}/eval-runs/${runId}`, opts)).json()
    );
  }

  async getResults(runId: number, opts?: { signal?: AbortSignal }): Promise<EvalResult[]> {
    return (await this.http.fetchOk(`${this.http.getBaseUrl()}/eval-runs/${runId}/results`, opts)).json();
  }

  // ── Project-level cross-script dashboard (sub-spec 1a) ────────────────────

  /**
   * One row per eval suite in the configured project, including the latest +
   * prior completed average score (drives the cross-script dashboard).
   */
  async listProjectSummaries(opts?: { signal?: AbortSignal }): Promise<EvalSuiteSummary[]> {
    return (await this.http.fetchOk(
      `${this.http.getBaseUrl()}/projects/${this.projectId}/eval-suite-summaries`,
      opts,
    )).json();
  }
}

File: crates/akribes-sdk/src/sub/evals.rs
use std::sync::Arc;

use crate::client::{AkribesClient, Inner};
use crate::error::Result;
use crate::models::*;

/// Sub-client for eval suites and runs. Obtained via [`AkribesClient::evals()`].
#[derive(Clone, Debug)]
pub struct EvalsClient {
    pub(crate) inner: Arc<Inner>,
    pub(crate) project_id: i64,
}

impl EvalsClient {
    pub(crate) fn new(inner: Arc<Inner>, project_id: i64) -> Self {
        Self { inner, project_id }
    }

    fn c(&self) -> AkribesClient {
        AkribesClient {
            inner: Arc::clone(&self.inner),
        }
    }

    fn project_url(&self) -> String {
        format!("{}/projects/{}", self.inner.base_url, self.project_id)
    }

    fn script_url(&self, name: &str) -> String {
        format!(
            "{}/scripts/{}",
            self.project_url(),
            urlencoding::encode(name)
        )
    }

    // ── Suites ──────────────────────────────────────────────────────────────

    /// List all eval suites for a script.
    pub async fn list_suites(&self, script_name: &str) -> Result<Vec<EvalSuite>> {
        let url = format!("{}/eval-suites", self.script_url(script_name));
        self.c().get_list(&url).await
    }

    /// Create a new eval suite.
    pub async fn create_suite(
        &self,
        script_name: &str,
        name: &str,
        runner_url: &str,
        config: Option<serde_json::Value>,
        auto_run_channels: Option<Vec<String>>,
    ) -> Result<EvalSuite> {
        let url = format!("{}/eval-suites", self.script_url(script_name));
        self.c()
            .post(
                &url,
                &CreateEvalSuiteRequest {
                    name,
                    runner_url,
                    config,
                    auto_run_channels,
                },
            )
            .await
    }

    /// Get an eval suite by ID.
    pub async fn get_suite(&self, script_name: &str, suite_id: i64) -> Result<Option<EvalSuite>> {
        let url = format!("{}/eval-suites/{}", self.script_url(script_name), suite_id);
        self.c().get_opt(&url).await
    }

    /// Update an eval suite (PATCH — only provided fields are changed).
    pub async fn update_suite(
        &self,
        script_name: &str,
        suite_id: i64,
        runner_url: Option<String>,
        config: Option<serde_json::Value>,
        auto_run_channels: Option<Vec<String>>,
    ) -> Result<EvalSuite> {
        let url = format!("{}/eval-suites/{}", self.script_url(script_name), suite_id);
        self.c()
            .patch(
                &url,
                &UpdateEvalSuiteRequest {
                    runner_url,
                    config,
                    auto_run_channels,
                },
            )
            .await
    }

    /// Delete an eval suite.
    pub async fn delete_suite(&self, script_name: &str, suite_id: i64) -> Result<bool> {
        let url = format!("{}/eval-suites/{}", self.script_url(script_name), suite_id);
        self.c().delete(&url).await
    }

    /// Check the health of a suite's eval runner.
    pub async fn check_runner_health(
        &self,
        script_name: &str,
        suite_id: i64,
    ) -> Result<serde_json::Value> {
        let url = format!(
            "{}/eval-suites/{}/health",
            self.script_url(script_name),
            suite_id
        );
        Ok(self
            .c()
            .get_opt::<serde_json::Value>(&url)
            .await?
            .unwrap_or(serde_json::json!({})))
    }

    // ── Trigger ─────────────────────────────────────────────────────────────

    /// Trigger an eval run for a suite.
    pub async fn trigger(
        &self,
        script_name: &str,
        suite_id: i64,
        source: Option<String>,
        channel: Option<String>,
        auto_publish: Option<bool>,
        triggered_by: Option<String>,
    ) -> Result<EvalRun> {
        let url = format!(
            "{}/eval-suites/{}/trigger",
            self.script_url(script_name),
            suite_id
        );
        self.c()
            .post(
                &url,
                &TriggerEvalRequest {
                    source,
                    channel,
                    auto_publish,
                    triggered_by,
                },
            )
            .await
    }

    // ── Runs & Results ──────────────────────────────────────────────────────

    /// List eval runs for a script, optionally filtered by suite.
    pub async fn list_runs(
        &self,
        script_name: &str,
        suite_id: Option<i64>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<EvalRun>> {
        #[derive(serde::Serialize)]
        struct Q {
            #[serde(skip_serializing_if = "Option::is_none")]
            suite_id: Option<i64>,
            #[serde(skip_serializing_if = "Option::is_none")]
            limit: Option<i64>,
            #[serde(skip_serializing_if = "Option::is_none")]
            offset: Option<i64>,
        }
        let base = format!("{}/eval-runs", self.script_url(script_name));
        let url = AkribesClient::url_with_query(
            &base,
            &Q {
                suite_id,
                limit,
                offset,
            },
        );
        self.c().get_list(&url).await
    }

    /// Get an eval run by ID (global endpoint, no project needed).
    pub async fn get_run(&self, run_id: i64) -> Result<Option<EvalRun>> {
        let url = format!("{}/eval-runs/{}", self.inner.base_url, run_id);
        self.c().get_opt(&url).await
    }

    /// Cancel a pending or running eval run. Returns the updated run.
    pub async fn cancel(&self, run_id: i64) -> Result<EvalRun> {
        let url = format!("{}/eval-runs/{}", self.inner.base_url, run_id);
        self.c().delete_json(&url).await
    }

    /// Get results for an eval run.
    pub async fn get_results(&self, run_id: i64) -> Result<Vec<EvalResult>> {
        let url = format!("{}/eval-runs/{}/results", self.inner.base_url, run_id);
        self.c().get_list(&url).await
    }

    // ── Project-level cross-script dashboard (sub-spec 1a) ──────────────────

    /// One summary row per eval suite in this project, with the latest +
    /// prior-completed average score (drives the cross-script dashboard).
    pub async fn list_project_summaries(&self) -> Result<Vec<EvalSuiteSummary>> {
        let url = format!("{}/eval-suite-summaries", self.project_url());
        self.c().get_list(&url).await
    }
}
